### PR TITLE
Improve integration of Linux package build with build pipelines

### PR DESCRIPTION
**Pull request by aahlenst**
During the last release, multiple problems surfaced:

* Mismatch of architecture names used by build pipelines and Gradle build caused build failures because not all Jenkins jobs have been updated.
* The Jenkins job that is part of the build pipeline pushed the releases into the repositories for the nightly builds because of misconfiguration.
* The Jenkins job that is part of the build pipeline failed if only a JDK but no JRE was available.
* The Jenkins job that actually pushes the releases frequently fails because it tries to re-release artifacts.

This pull request should address all these problems:

* Gradle build uses the same architecture signifiers as Jenkins build pipelines.
* Gradle build does not try to re-release existing artifacts, only updates metadata if necessary.
* Two shell scripts have been added that should replace the scripts in the Jenkins jobs. `create_installer_linux.sh` is for the regular build pipeline and creates installers from the tarballs copied into the Jenkins workspace. `standalone_create_installer_linux.sh` is a replacement for `artifactory_release_publisher` and `artifactory_release_publisher-bot`. As `standalone_create_installer_windows`, it downloads official releases from adoptopenjdk.net and repackages those using `create_installer_linux.sh`. Both shell scripts can cope with missing releases.

### CODE REVIEW

**Comment by karianna** [comment]
@gdams and @johnoliver Can I get you to take a look - this will obviously be valuable if added with care.

@aahlenst Thanks for the PR - it looks sane to me although is there some tidy-up of old scripts and/or Jenkins jobs required as part of this?  Would be good to roughly document that.

**Comment by aahlenst** [comment]
`artifactory_release_publisher-bot` and `artifactory_release_publisher` can be disabled and dropped after we have verified that the replacement is okay. `create_installer_linux.sh` should be a drop-in replacement for the job with the same name. The existing script that is part of the Jenkins job can be removed. Invoke `create_installer_linux.sh` instead. I can prepare and test the standalone job if someone would create it for me.

Does that suffice?

**Comment by karianna** [comment]
> `artifactory_release_publisher-bot` and `artifactory_release_publisher` can be disabled and dropped after we have verified that the replacement is okay. `create_installer_linux.sh` should be a drop-in replacement for the job with the same name. The existing script that is part of the Jenkins job can be removed. Invoke `create_installer_linux.sh` instead. I can prepare and test the standalone job if someone would create it for me.
> 
> Does that suffice?

Sounds good to me.  Let me see what GitHub group you're part of and you can create your own job

**Comment by karianna** [comment]
OK I'm blocked on a Jenkins 403 issue (we need to fix some Ngnix config) - I'll revisit this on Monday after we make that fix

**Comment by aahlenst** [comment]
@karianna Could you already look into the Jenkins/nginx issue?

File: linux/build.gradle [hunk 1]
```diff
@@ -86,8 +86,11 @@ def getJdkMajorVersion() {
         throw new IllegalArgumentException("Unparsable JDK major version '$versionString'")
     }
     int version = Integer.valueOf(versionString)
-    if (version < 8) {
-        throw new IllegalArgumentException("JDK major version '$version' out of range")
+
+    // Do not change the numbers without updating the metadata declarations
+    // (dependencies, jinfo priorities)!
+    if (version < 8 || version > 12) {
```
Comment (karianna) [inline linux/build.gradle]: Should we be going to 13 now?

File: linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadDebianPackage.java [hunk 2]
```diff
@@ -46,10 +57,44 @@ protected void uploadPackages(ArtifactoryCredentials artifactoryCredentials) {
...
+        }
+
+        for (RepoPath searchResult : searchResults) {
+            if (Objects.equals(getRepository(), searchResult.getRepoKey())
+                    && Objects.equals(this.targetFilePath(), searchResult.getItemPath())) {
+                return true;
+            }
+        }
+
+        return false;
```
Comment (karianna) [inline linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadDebianPackage.java]: Should we log this case or is it common?

File: linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadRpmPackage.java [hunk 3]
```diff
@@ -50,14 +59,37 @@ protected void uploadPackages(ArtifactoryCredentials artifactoryCredentials) {
...
+                .artifactsByName(getPackageToPublish().getName())
+                .doSearch();
+
+        if (searchResults.isEmpty()) {
+            return false;
+        }
+
+        for (RepoPath searchResult : searchResults) {
+            if (Objects.equals(getRepository(), searchResult.getRepoKey())
+                    && Objects.equals(remotePath.substring(1), searchResult.getItemPath())) {
```
Comment (karianna) [inline linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadRpmPackage.java]: slight preference to pull `remotePath.substring(1)` into a named variable,  but no biggy.

File: linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadRpmPackage.java [hunk 4]
```diff
@@ -50,14 +59,37 @@ protected void uploadPackages(ArtifactoryCredentials artifactoryCredentials) {
...
+                return true;
             }
         }
+
+        return false;
```
Comment (karianna) [inline linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadRpmPackage.java]: Should we log this?

File: linux/create_installer_linux.sh [hunk 5]
```diff
@@ -0,0 +1,135 @@
+#! /bin/bash
```
Comment (karianna) [inline linux/create_installer_linux.sh]: Has this whole file had shellcheck run on it?

File: linux/create_installer_linux.sh [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+finish() {
+    if [ -d "$DISTRIBUTION_DIR" ] ; then
+        rm -rf "$DISTRIBUTION_DIR"
+    fi
+}
+trap finish EXIT
+
+echoerr() { printf "%s\n" "$*" >&2; }
+
+case $RELEASE_TYPE in
```
Comment (karianna) [inline linux/create_installer_linux.sh]: could this block be wrapped in a function?

File: linux/create_installer_linux.sh [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        ARTIFACTORY_REPOSITORY_RPM=""
+        GRADLE_TASK="build"
+        ;;
+    *)
+        echoerr "Unsupported build type"
+        exit 0
+        ;;
+esac
+
+JDK_TARBALLS=( "$WORKSPACE"/OpenJDK*.tar.gz )
```
Comment (karianna) [inline linux/create_installer_linux.sh]: could this block be wrapped in a function?

File: linux/build.gradle (hunk 1 above)
Comment (aahlenst) [reply linux/build.gradle]: It‘s prepared but comes in a separate pull request.

File: linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadDebianPackage.java (hunk 2 above)
Comment (aahlenst) [reply linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadDebianPackage.java]: It‘s logged, line 31.

File: linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadRpmPackage.java (hunk 4 above)
Comment (aahlenst) [reply linux/buildSrc/src/main/java/net/adoptopenjdk/installer/UploadRpmPackage.java]: See line 36.

File: linux/create_installer_linux.sh (hunk 6 above)
Comment (aahlenst) [reply linux/create_installer_linux.sh]: What would be the advantage? As far as I know, Bash does not really have return values and the code isn‘t re-used anywhere. Happy to make the change, but hard if I don‘t see what you are aiming at.

Comment (karianna) [reply linux/create_installer_linux.sh]: In that case, we can just add a comment at the top of the block explaining what it broadly does

Comment (karianna) [reply linux/create_installer_linux.sh]: Hmm, the previous comment didn't save.

In that case, perhaps add a comment at the top of the block?

File: linux/create_installer_linux.sh (hunk 5 above)
Comment (aahlenst) [reply linux/create_installer_linux.sh]: Yes, but I missed one issue. It's fixed now.

File: linux/create_installer_linux.sh (hunk 6 above)
Comment (aahlenst) [reply linux/create_installer_linux.sh]: Comments have been added.

**Pull request merged**